p2p/net/mock/mock_stream.go

```diff
@@ -1,20 +1,41 @@
 package mocknet
 
 import (
-	"io"
-
+	"bytes"
+	"fmt"
 	inet "github.com/ipfs/go-ipfs/p2p/net"
+	"io"
+	"time"
 )
 
 // stream implements inet.Stream
 type stream struct {
 	io.Reader
 	io.Writer
-	conn *conn
+	conn      *conn
+	toDeliver chan *transportObject
+	done      chan bool
+}
+
+type transportObject struct {
+	msg         []byte
+	arrivalTime time.Time
+}
+
+//  How to handle errors with writes?
+func (s *stream) Write(p []byte) (n int, err error) {
+	l := s.conn.link
+	delay := l.GetLatency() + l.RateLimit(len(p))
+	t := time.Now().Add(delay)
+	s.toDeliver <- &transportObject{msg: p, arrivalTime: t}
+	return len(p), nil
 }
 
 func (s *stream) Close() error {
 	s.conn.removeStream(s)
+	close(s.toDeliver)
```

Comment:
- closing `s.toDeliver` here is not safe with the send in `Write` above (`s.toDeliver <- ...`).
- `s.Close()` needs to send a signal to the writes to error out _before_ closing this channel. (it's possible this channel doesn't even need to be closed.